core/order.go

```diff
@@ -252,6 +292,237 @@ func (n *OpenBazaarNode) Purchase(data *PurchaseData) error {
+						shipsToMe = true
+						break
+					}
+				}
+				if !shipsToMe {
+					return 0, errors.New("Listing does ship to selected country")
+				}
+
+				// Check service exists
+				var service *pb.Listing_ShippingOption_Service
+				for _, shippingService := range option.Services {
+					if strings.ToLower(shippingService.Name) == strings.ToLower(item.ShippingOption.Service) {
+						service = shippingService
+					}
+				}
+				if service == nil {
+					return 0, errors.New("Shipping service not found in listing")
+				}
+				shippingSatoshi, err := n.getPriceInSatoshi(service.Price)
+				if err != nil {
+					return 0, err
+				}
+				shippingPrice := uint64(item.Quantity) * shippingSatoshi
+				itemShipping += shippingPrice
+
+				// Apply shipping rules
+				if option.ShippingRules != nil {
+					for _, rule := range option.ShippingRules.Rules {
+						if int(option.ShippingRules.RuleType) == 0 {
+							if item.Quantity >= rule.MinRange && item.Quantity <= rule.MaxRange {
+								rulePrice, err := n.getPriceInSatoshi(rule.Price)
+								if err != nil {
+									return 0, err
+								}
+								itemShipping -= rulePrice
+							}
+						} else if int(option.ShippingRules.RuleType) == 1 {
+							if item.Quantity >= rule.MinRange && item.Quantity <= rule.MaxRange {
+								itemShipping -= shippingPrice
+								rulePrice, err := n.getPriceInSatoshi(rule.Price)
+								if err != nil {
+									return 0, err
+								}
+								itemShipping += rulePrice
+							}
+						} else if int(option.ShippingRules.RuleType) == 2 {
+							weight := listing.Item.Grams * float32(item.Quantity)
+							if uint32(weight) >= rule.MinRange && uint32(weight) <= rule.MaxRange {
+								itemShipping -= shippingPrice
+								rulePrice, err := n.getPriceInSatoshi(rule.Price)
+								if err != nil {
+									return 0, err
+								}
+								itemShipping += rulePrice
+							}
+						} else if int(option.ShippingRules.RuleType) == 3 {
+							itemShipping -= shippingPrice
+							rulePrice, err := n.getPriceInSatoshi(rule.Price)
+							if err != nil {
+								return 0, err
+							}
+							cs := combinedShipping{
+								quantity: int(item.Quantity),
+								price:    shippingSatoshi,
+								add:      true,
+								modifier: rulePrice,
+							}
+							combinedOptions = append(combinedOptions, cs)
+
+						} else if int(option.ShippingRules.RuleType) == 4 {
+							itemShipping -= shippingPrice
+							rulePrice, err := n.getPriceInSatoshi(rule.Price)
+							if err != nil {
+								return 0, err
+							}
+							cs := combinedShipping{
+								quantity: int(item.Quantity),
+								price:    shippingSatoshi,
+								add:      false,
+								modifier: rulePrice,
+							}
+							combinedOptions = append(combinedOptions, cs)
+						}
+					}
+				}
+				shippingTotal += itemShipping
+			}
+		}
+	}
+
+	// Process combined shipping rules
+	if len(combinedOptions) > 0 {
+		lowestPrice := uint64(2100000000000000)
```

Comment:
This seems like it should be a const even though it's fairly clear what it represents.